identity/identity.go

```diff
@@ -103,66 +99,91 @@ func GetIdentityHeader(ctx context.Context) string {
 }
 
-func checkHeader(id *XRHID, w http.ResponseWriter) error {
+var (
+	MissingIdentityError      = errors.New("missing x-rh-identity header")
+	DecodeIdentityError       = errors.New("unable to b64 decode x-rh-identity header")
+	InvalidOrgIdIdentityError = errors.New("x-rh-identity header has an invalid or missing org_id")
+	MissingIdentityTypeError  = errors.New("x-rh-identity header is missing type")
+)
 
+func checkHeader(id *XRHID) error {
 	if (id.Identity.Type == "Associate" || id.Identity.Type == "X509") && id.Identity.AccountNumber == "" {
 		return nil
 	}
 
 	if id.Identity.OrgID == "" && id.Identity.Internal.OrgID == "" {
-		return doError(w, 400, "x-rh-identity header has an invalid or missing org_id")
+		return InvalidOrgIdIdentityError
 	}
 
 	if id.Identity.Type == "" {
-		return doError(w, 400, "x-rh-identity header is missing type")
+		return MissingIdentityTypeError
 	}
 
 	return nil
 }
 
-// EnforceIdentity extracts the X-Rh-Identity header and places the contents into the
-// request context.  If the Identity is invalid, the request will be aborted.
-func EnforceIdentity(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		rawHeaders := r.Header["X-Rh-Identity"]
+// if org_id is not defined at the top level, use the internal one
+// https://issues.redhat.com/browse/RHCLOUD-17717
+func topLevelOrgIDFallback(identity *XRHID) {
+	if identity.Identity.OrgID == "" && identity.Identity.Internal.OrgID != "" {
+		identity.Identity.OrgID = identity.Identity.Internal.OrgID
+	}
+}
 
-		// must have an x-rh-id header
-		if len(rawHeaders) != 1 {
-			doError(w, 400, "missing x-rh-identity header")
-			return
-		}
+// DecodeIdentity returns identity value decoded from a base64 JSON encoded
+// string. The function performs series of checks and will return errors
+// for invalid identities.
+//
+// To put identity into a context, use With or DecodeIdentityCtx functions.
+func DecodeIdentity(header string) (XRHID, error) {
+	if header == "" {
+		return XRHID{}, MissingIdentityError
+	}
 
-		// must be able to base64 decode header
-		idRaw, err := base64.StdEncoding.DecodeString(rawHeaders[0])
-		if err != nil {
-			doError(w, 400, "unable to b64 decode x-rh-identity header")
-			return
-		}
+	// must be able to base64 decode header
+	idRaw, err := base64.StdEncoding.DecodeString(header)
+	if err != nil {
+		return XRHID{}, DecodeIdentityError
+	}
 
-		var jsonData XRHID
-		err = json.Unmarshal(idRaw, &jsonData)
-		if err != nil {
-			log.Printf("unable to unmarshal x-rh-identity header: %s", err)
-			doError(w, 400, "x-rh-identity header does not contain valid JSON")
-			return
-		}
+	var xRhId XRHID
+	err = json.Unmarshal(idRaw, &xRhId)
+	if err != nil {
+		return XRHID{}, fmt.Errorf("x-rh-identity header does not contain valid JSON: %w", err)
+	}
+
+	topLevelOrgIDFallback(&xRhId)
+
+	err = checkHeader(&xRhId)
+	if err != nil {
+		return XRHID{}, err
+	}
+
+	return xRhId, nil
+}
```

Comment:
This was extracted from the middleware function, instead of writing to HTTP body it now returns error type.